src/slice.rs

```diff
@@ -842,15 +842,11 @@ macro_rules! s(
         }
     };
     // empty call, i.e. `s![]`
-    (@parse ::core::marker::PhantomData::<$crate::Ix0>, ::core::marker::PhantomData::<$crate::Ix0>, []) => {
+    (@parse $in_dim:expr, $out_dim:expr, []) => {
         {
             #[allow(unsafe_code)]
             unsafe {
-                $crate::SliceInfo::new_unchecked(
-                    [],
-                    ::core::marker::PhantomData::<$crate::Ix0>,
-                    ::core::marker::PhantomData::<$crate::Ix0>,
-                )
+                $crate::SliceInfo::new_unchecked([], $in_dim, $out_dim)
```

Comment:
what's the impact on the `unsafe` code block here?

Reply:
It's still necessary, as `SliceInfo::new_unchecked` is unsafe.

Technically, this means that someone using the macro could write

```rust
s! { @parse
{
    // statements in an unsafe context
    ::core::marker::PhantomData::<::ndarray::Ix0>
},
::core::marker::PhantomData<::ndaray::Ix0>,
[]
}
```

to get an `unsafe` context without writing `unsafe`, but this is an internal rule, so that's not too worrisome. However, it is reasonable to evaluate the arguments outside of the `unsafe` block just for thoroughness.

Reply:
Thanks for bringing this up. Some of the other `@parse` arms had safety issues if called directly. I've created #1196 to fix that. With that change, we can change this PR to modify the "empty case" arm to the following:

```rust
    // empty call, i.e. `s![]`
    (@parse $in_dim:expr, $out_dim:expr, []) => {
        {
            debug_assert_eq!($in_dim, ::core::marker::PhantomData::<$crate::Ix0>);
            debug_assert_eq!($out_dim, ::core::marker::PhantomData::<$crate::Ix0>);
            (
                [],
                ::core::marker::PhantomData::<$crate::Ix0>,
                ::core::marker::PhantomData::<$crate::Ix0>,
            )
        }
    };
```

This is safe, since it always expands to valid values for the `SliceInfo::new_unchecked` call (which is in the `($($t:tt)*)` arm with #1196). It has debug assertions to ensure that `$in_dim` and `$out_dim` are the expected values. (They'll always be the expected values, unless the user calls `s![@parse ...]` directly.